File: bit-manipulation/1356.py
# Amazon


# Approach-1
# Tc = O(sorting + 1 bit count)
# Tc = O(n.logn + log(num)) [n for loop, logn for sorting, log(num) -> for 1 bit count in any number]
class Solution(object):
    def countOneBit(self, num):
        count = 0
        while num != 0: # log^2(num)
            count += num & 1
            num >>= 1
        return count
    def sortByBits(self, arr): #n.logn
        arr.sort(key = lambda x: (self.countOneBit(x), x))
        return arr
        


# Approach-2(without using lambda function)
# Tc = O(n.logn) [only for sorting and loop]
class Solution(object):
    def sortByBits(self, arr):
        def countOneBit(x):
            return bin(x).count('1')
        arr.sort(key = lambda x: (countOneBit(x), x))
        return arr

File: bit-manipulation/1356.cpp
class Solution {
public:
    int countOneBit(int num){
        int count = 0;
        while(num != 0){
            count += num & 1;
            num >>= 1;
        }
        return count;
    }
    vector<int> sortByBits(vector<int>& arr) {
        auto lambda= [&](int &a, int&b){
            int CA = countOneBit(a);
            int CB = countOneBit(b);
            if(CA == CB)
                return a < b;
            return CA < CB;
        };
        sort(begin(arr), end(arr), lambda);
        return arr;
    }
};



// Tc = O(n.logn)
class Solution {
public:
    vector<int> sortByBits(vector<int>& arr) {
        auto lambda= [&](int &a, int&b){
            int CA = __builtin_popcount(a);
            int CB = __builtin_popcount(b);
            if(CA == CB)
                return a < b;
            return CA < CB;
        };
        sort(begin(arr), end(arr), lambda);
        return arr;
    }
};